rllib/core/optim/tests/test_rl_optimizer_tf.py

```diff
@@ -0,0 +1,477 @@
+        """Compute a loss"""
+        action_dist = fwd_out["action_dist"]
+        loss = -tf.math.reduce_mean(action_dist.log_prob(batch[SampleBatch.ACTIONS]))
+        loss_dict = {
+            "total_loss": loss,
+        }
+        return loss_dict
+
+    def get_state(self):
+        return {
+            key: optim.get_weights() for key, optim in self.get_optimizers().items()
+        }
+
+    def set_state(self, state: Mapping[str, Any]) -> None:
+        assert set(state.keys()) == set(self.get_state().keys()) or not state
+        for key, optim_dict in state.items():
+            self.get_optimizers()[key].set_weights(optim_dict)
+
+    def _configure_optimizers(self) -> Mapping[str, Any]:
+        return {
+            "policy": tf.keras.optimizers.Adam(
+                learning_rate=self._config.get("lr", 1e-3)
+            )
+        }
+
+
+class BCTFTrainer:
+    def __init__(self, config: Mapping[str, any]) -> None:
+        module_config = config["module_config"]
+        optimizer_config = {}
+        self._module = BCTFModule(module_config)
+        self._rl_optimizer = BCTFOptimizer(self._module, optimizer_config)
+
+    @staticmethod
+    def on_after_compute_gradients(
+        gradients_dict: Mapping[str, Any]
+    ) -> Mapping[str, Any]:
+        """Called after gradients have been computed.
+
+        Args:
+            gradients_dict (Mapping[str, Any]): A dictionary of gradients.
+
+        Note the relative order of operations looks like this:
+            fwd_out = forward_train(batch)
+            loss = compute_loss(batch, fwd_out)
+            gradients = compute_gradients(loss)
+            ---> post_processed_gradients = on_after_compute_gradients(gradients)
+            apply_gradients(post_processed_gradients)
+
+        Returns:
+            Mapping[str, Any]: A dictionary of gradients.
+        """
+        return gradients_dict
+
+    def compile_results(
+        self,
+        batch: NestedDict,
+        fwd_out: Mapping[str, Any],
+        postprocessed_loss: Mapping[str, Any],
+        post_processed_gradients: Mapping[str, Any],
+    ) -> Mapping[str, Any]:
+        """Compile results from the update.
+
+        Args:
+            batch: The batch that was used for the update.
+            fwd_out: The output of the forward train pass.
+            postprocessed_loss: The loss after postprocessing.
+            post_processed_gradients: The gradients after postprocessing.
+
+        Returns:
+            A dictionary of results.
+        """
+
+        loss_numpy = convert_to_numpy(postprocessed_loss)
+        results = {
+            "avg_reward": tf.math.reduce_mean(batch["rewards"]).numpy(),
+        }
+        results.update(loss_numpy)
+        return results
+
+    @tf.function
+    def _do_update_fn(self, batch: SampleBatch) -> Mapping[str, Any]:
+        with tf.GradientTape() as tape:
+            # fwd_out = self._module.forward_train(batch)
```

Comment:
We should call forward_train or call the module directly? I feel like what ever solution we opt for here should also be applied to the torch trainer.